tool/tsh/db.go

```diff
@@ -869,6 +881,71 @@ func isMFADatabaseAccessRequired(cf *CLIConf, tc *client.TeleportClient, databas
 	return mfaResp.GetRequired(), nil
 }
 
+// checkRoute checks if a database route will be denied access.
+// Standard RBAC checks are still enforced
+// at connection time - this check just attempts to "fail fast" with
+// a user-friendly error.
+func checkRoute(tc *client.TeleportClient, cf *CLIConf, profile *client.ProfileStatus, dbRoute *tlsca.RouteToDatabase, db types.Database) (err error) {
+	roleSet, err := fetchCurrentClusterRoleSet(tc, cf, profile)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	dbUsers := roleSet.EnumerateDatabaseUsers(db)
+	// catch the cases where a user will be denied no matter what.
+	if dbUsers.WildcardDenied() {
+		return trace.AccessDenied("Your Teleport role(s) deny login as any db user (db_users wildcard %q is denied). Ask your Teleport administrator to grant you access. See: https://goteleport.com/docs/database-access/rbac/", types.Wildcard)
+	}
+	if !dbUsers.IsAnyAllowed() {
+		return trace.AccessDenied("Your Teleport role(s) do not allow login as any db user for database %q. Ask your Teleport administrator to grant you access. See: https://goteleport.com/docs/database-access/rbac/", dbRoute.ServiceName)
+	}
+
+	// if the user asked for a specific --db-user we can check it here
+	if dbRoute.Username != "" && !dbUsers.IsAllowed(dbRoute.Username) {
+		return trace.AccessDenied("Your Teleport role(s) do not allow login to database %q as db user %q. Use `tsh db ls` to see available logins or ask your Teleport administrator to grant you access. See: https://goteleport.com/docs/database-access/rbac/", dbRoute.ServiceName, dbRoute.Username)
+	}
+
+	// we only enforce db_names access for postgres and mongo, so check access for those protocols
+	switch dbRoute.Protocol {
+	case defaults.ProtocolPostgres, defaults.ProtocolMongoDB:
+		dbNames := roleSet.EnumerateDatabaseNames(db)
+		// catch the cases where a user will be denied no matter what.
+		if dbNames.WildcardDenied() {
+			return trace.AccessDenied("Your Teleport role(s) deny db_names wildcard %q (db_names RBAC is used for %q protocol). Ask your Teleport administrator to grant you access. See: https://goteleport.com/docs/database-access/rbac/", types.Wildcard, dbRoute.Protocol)
```

Comment:
The "(db_names RBAC is used for %q protocol)" seems to  superfluous form the user perspective.